File: codejam/qual2020/ppr/ppr.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

bool has_overlap(vector<int> a, vector<int> b)
{
  if (a.empty() || b.empty()) {
    return false;
  }
  return max(a[0], b[0]) < min(a[1], b[1]);
}

bool has_overlap_with(vector<vector<int> >& activities, vector<int> activity)
{
  if (activities.empty()) {
    return false;
  }
  for (int i = 0; i < activities.size(); ++i) {
    if (has_overlap(activities[i], activity)) {
      return true;
    }
  }
  return false;
}

string assign_activities(vector<vector<int> > activities)
{
  int n = activities.size();
  vector<string> as(n);
  vector<vector<int> > cs(n);
  vector<vector<int> > js(n);
  sort(activities.begin(), activities.end());

  for (int i = 0; i < n; ++i) {
    vector<int> activity = activities[i];
    if (has_overlap_with(cs, activity)) {
      if (has_overlap_with(js, activity)) {
        return "IMPOSSIBLE";
      }
      js.push_back(activity);
      as[activity[2]] = "J";
    } else {
      cs.push_back(activity);
      as[activity[2]] = "C";
    }
  }
  string s = "";
  for (int i = 0; i < n; ++i) {
    s += as[i];
  }
  return s;
}

void test_case()
{
  int n;
  scanf("%d", &n);
  vector<vector<int> > activities(n);
  for (int i = 0; i < n; ++i) {
    vector<int> activity(3);
    scanf("%d", &activity[0]);
    scanf("%d", &activity[1]);
    activity[2] = i;
    activities[i] = activity;
  }
  cout << assign_activities(activities) << endl;
}

int main()
{
  int T;
  scanf("%d", &T);
  for (int i = 1; i <= T; ++i) {
    printf("Case #%d: ", i);
    test_case();
  }
}

File: codejam/qual2020/ppr/ppr.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func readInput() []string {
	lines := []string{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

func logCase(num int, out string) {
	fmt.Printf("Case #%d: %s\n", num, out)
}

type Interval struct {
	start, end int
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func (i Interval) hasOverlap(i2 Interval) bool {
	return i.start == i2.start || i.end == i2.end ||
		(i.start < i2.start && i.end > i2.start) ||
		(i2.start < i.start && i2.end > i.start)
}

type Schedule []Interval

func (s Schedule) hasOverlap(i Interval) bool {
	for _, sc := range s {
		if sc.hasOverlap(i) {
			return true
		}
	}
	return false
}

type Case struct {
	n        int
	schedule Schedule
}

func (c Case) assignActivities() string {
	as := ""
	cs := Schedule{}
	js := Schedule{}
	for _, s := range c.schedule {
		if cs.hasOverlap(s) {
			if js.hasOverlap(s) {
				return "IMPOSSIBLE"
			}
			js = append(js, s)
			as += "J"
		} else {
			cs = append(cs, s)
			as += "C"
		}
	}
	return as
}

func (c Case) String() string {
	return c.assignActivities()
}

type Input struct {
	t     int
	cases []Case
}

func parseInput(lines []string) Input {
	r := Input{}
	r.t, _ = strconv.Atoi(lines[0])
	for i := 0; i < len(lines)-1; i++ {
		c := Case{}
		c.n, _ = strconv.Atoi(lines[i+1])
		for j := 0; j < c.n; j++ {
			is := lines[i+2]
			isa := strings.Split(is, " ")
			interval := Interval{}
			interval.start, _ = strconv.Atoi(isa[0])
			interval.end, _ = strconv.Atoi(isa[1])
			c.schedule = append(c.schedule, interval)
			i++
		}
		r.cases = append(r.cases, c)
	}
	return r
}

func solve(c Case) string {
	return fmt.Sprintf("%s", c)
}

func main() {
	in := parseInput(readInput())
	for i, c := range in.cases {
		logCase(i+1, solve(c))
	}
}
